test_conformance/api/test_api_consistency.cpp

```diff
@@ -473,14 +437,689 @@ int test_consistency_pipes(cl_device_id deviceID, cl_context context,
+            "but clGetSupportedImageFormats(CL_MEM_KERNEL_READ_AND_WRITE) "
+            "returned a non-empty set");
+    }
+    else
+    {
+        test_assert_error(
+            totalReadWriteImageFormats != 0,
+            "CL_DEVICE_MAX_READ_WRITE_IMAGE_ARGS is non-zero "
+            "but clGetSupportedImageFormats(CL_MEM_KERNEL_READ_AND_WRITE) "
+            "returned an empty set");
+    }
+
+    return TEST_PASS;
+}
+
+int test_consistency_2d_image_from_buffer(cl_device_id deviceID,
+                                          cl_context context,
+                                          cl_command_queue queue,
+                                          int num_elements)
+{
+    // clGetDeviceInfo, passing CL_DEVICE_IMAGE_PITCH_ALIGNMENT or
+    // CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT
+    // May return 0, indicating that device does not support Creating a 2D Image
+    // from a Buffer.
+    cl_int error;
+
+    const cl_image_format imageFormat = { CL_RGBA, CL_UNORM_INT8 };
+    const size_t imageDim = 2;
+    const size_t elementSize = 4;
+    const size_t bufferSize = imageDim * imageDim * elementSize;
+
+    clMemWrapper buffer;
+    clMemWrapper image;
+
+    cl_uint imagePitchAlignment = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_IMAGE_PITCH_ALIGNMENT,
+                            sizeof(imagePitchAlignment), &imagePitchAlignment,
+                            NULL);
+    test_error(error,
+               "Unable to query "
+               "CL_DEVICE_IMAGE_PITCH_ALIGNMENT");
+
+    cl_uint imageBaseAddressAlignment = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT,
+                            sizeof(imageBaseAddressAlignment),
+                            &imageBaseAddressAlignment, NULL);
+    test_error(error,
+               "Unable to query "
+               "CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT");
+
+    bool supports_cl_khr_image2d_from_buffer =
+        is_extension_available(deviceID, "cl_khr_image2d_from_buffer");
+
+    if (imagePitchAlignment == 0 || imageBaseAddressAlignment == 0)
+    {
+        // This probably means that Creating a 2D Image from a Buffer is not
+        // supported.
+
+        // Test setup:
+        buffer =
+            clCreateBuffer(context, CL_MEM_READ_ONLY, bufferSize, NULL, &error);
+        test_error(error, "Unable to create test buffer");
+
+        // Check that both queries return zero:
+        test_assert_error(imagePitchAlignment == 0,
+                          "CL_DEVICE_IMAGE_PITCH_ALIGNMENT returned a non-zero "
+                          "value but CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT "
+                          "returned 0");
+        test_assert_error(imagePitchAlignment == 0,
+                          "CL_DEVICE_IMAGE_PITCH_ALIGNMENT returned 0 but "
+                          "CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT returned a "
+                          "non-zero value");
```

Comment:
These will never fire.  Given the messages, I'm guessing the conditions are incomplete.

Reply:
No, these may fire, since we'll enter the conditional if _either_ of the conditions are true.  These checks are to ensure that if either condition is true then both conditions are true.

Reply:
Yup, I read the code a bit too quickly. The second check's condition should probably be `imageBaseAddressAlignment == 0` though.

Reply:
Yes it should be `imageBaseAddressAlignment` - fixed!